src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Logging/ProjectDiagnosticOutputServiceExtensions.cs

```diff
@@ -91,15 +51,18 @@ public static void WriteLine(this IProjectLogger logger, string format, object?
         /// <exception cref="FormatException">
         ///     The format specification in <paramref name="format"/> is invalid.
         /// </exception>
-        public static void WriteLine(this IProjectLogger logger, string format, object? argument1, object? argument2)
+        public static void WriteLine(this IProjectDiagnosticOutputService logger, string format, object? argument1, object? argument2)
         {
             Requires.NotNull(logger, nameof(logger));
 
-            logger.WriteLine(new StringFormat(format, argument1, argument2));
+            if (logger.IsEnabled)
```

Comment:
>if (logger.IsEnabled) [](http://example.com/codeflow?start=12&length=21)

Should the IProjectDiagnosticOutputService be updated to check whether it is enabled before WriteLining? Seems odd to have to do this in the consumers.

Reply:
That service only accepts a `string`. Here we have format arguments and want to avoid allocating a string if the logger is disabled.

The format arguments here are `object` so we may still have some boxing. In places where we do larger amounts of logging, the caller of these extension methods may also check `IsEnabled`.

Reply:
Wouldn't having a WriteFormattedString in the logger and have the logger check whether it is enabled address these concerns?

Reply:
Potentially yes. In this case that'd be a change to the CPS API.

Though for performance it's simpler to have the client decide how it wants to handle it. For example, what API would that method have? Probably `WriteFormattedString(string format, params string[] args)` with a `params` array for the format args. That means that the array of arguments is allocated (and any struct args boxed) on every call, even if the logging is disabled.

Ideally we want "you don't pay for what you don't use" which is hard to achieve for this kind of logging with today's core types. I'd hoped that `FormattableString` might be able to help here somehow, but that is still a class and has its own limitations. In the .NET Project System we have [`StringFormat`](https://github.com/dotnet/project-system/blob/main/src/Microsoft.VisualStudio.ProjectSystem.Managed/StringFormat.cs) as an attempt to avoid allocation of params arrays, but it's not a perfect pattern either, and we keep it `internal`.